sdk/program/src/message/compiled_keys.rs

```diff
@@ -85,8 +94,69 @@ impl CompiledKeys {
             .chain(self.readonly_non_signer_keys)
             .collect();
 
-        Some((header, static_account_keys))
+        Ok((header, static_account_keys))
+    }
+
+    pub(crate) fn try_extract_table_lookup(
+        &mut self,
+        lookup_table_account: &AddressLookupTableAccount,
+    ) -> Result<Option<(MessageAddressTableLookup, LoadedAddresses)>, CompileError> {
+        let (writable_indexes, drained_writable_keys) = try_drain_keys_found_in_lookup_table(
+            &mut self.writable_non_signer_keys,
+            &lookup_table_account.addresses,
+        )?;
+        let (readonly_indexes, drained_readonly_keys) = try_drain_keys_found_in_lookup_table(
+            &mut self.readonly_non_signer_keys,
+            &lookup_table_account.addresses,
+        )?;
+
+        // Don't extract lookup if no keys were found
+        if writable_indexes.is_empty() && readonly_indexes.is_empty() {
+            return Ok(None);
+        }
+
+        Ok(Some((
+            MessageAddressTableLookup {
+                account_key: lookup_table_account.key,
+                writable_indexes,
+                readonly_indexes,
+            },
+            LoadedAddresses {
+                writable: drained_writable_keys,
+                readonly: drained_readonly_keys,
+            },
+        )))
+    }
+}
+
+fn try_drain_keys_found_in_lookup_table(
+    keys: &mut Vec<Pubkey>,
+    lookup_table_addresses: &[Pubkey],
+) -> Result<(Vec<u8>, Vec<Pubkey>), CompileError> {
+    let mut lookup_table_indexes = Vec::new();
+    let mut drained_keys = Vec::new();
+    let mut i = 0;
+    while i < keys.len() {
```

Comment:
I don't think we can avoid `Vec::remove` here even in the reverse direction so I think the benefit of doing the removal in reverse is negligible. I'd also like to keep the account keys roughly ordered by pubkey and removing in reverse would the reverse the order of lookup table indexes.

Reply:
the idea was to avoid the shift, not the `Vec::remove`. need [`drain_filter`](https://doc.rust-lang.org/beta/std/vec/struct.Vec.html#method.drain_filter)

Reply:
My implementation is actually from the `drain_filter` docs where it says "Using this method is equivalent to the following code"